`GetExecutorRequestResource`  panics when dynamic allocation is enabled with batch scheduler

### What happened?

- [x] ✋ I have searched the open/closed issues and my issue is not listed.


Controller panics at https://github.com/kubeflow/spark-operator/blob/2e96ece15939cd7eb365b34268f416ae90f67ec9/pkg/util/sparkapplication.go#L469
 when a SparkApplication uses dynamic allocation with a batch scheduler (example Volcano), line 469 dereferences `*app.Spec.Executor.Instances` without nil check. When dynamic allocation is enabled, `Instances` remains nil because the defaulting logic at 
https://github.com/kubeflow/spark-operator/blob/2e96ece15939cd7eb365b34268f416ae90f67ec9/api/v1beta2/defaults.go#L95-L98
 only sets it when dynamic allocation is disabled.

### Reproduction Code

1. Install spark-operator with `--set controller.batchScheduler.enable=true`
2. Install Volcano scheduler
3. Apply this SparkApplication
```yaml
apiVersion: sparkoperator.k8s.io/v1beta2
kind: SparkApplication
metadata:
  name: test-dynamic-alloc
spec:
  type: Scala
  mode: cluster
  image: spark:3.5.0
  mainClass: org.apache.spark.examples.SparkPi
  mainApplicationFile: local:///opt/spark/examples/jars/spark-examples.jar
  sparkVersion: "3.5.0"
  batchScheduler: volcano
  dynamicAllocation:
    enabled: true
    minExecutors: 1
    maxExecutors: 10
  driver:
    cores: 1
    memory: "512m"
    serviceAccount: spark
  executor:
    cores: 1
    memory: "512m"
```

### Expected behavior

controller handles nil `Executor.Instances` gracefully 

### Actual behavior

```bash
panic: runtime error: invalid memory address or nil pointer dereference
github.com/kubeflow/spark-operator/v2/pkg/util.GetExecutorRequestResource
    /workspace/pkg/util/sparkapplication.go:469
```

### Environment & Versions

- Kubernetes Version: v1.30.0
- Spark Operator Version: v2.3.0
- Apache Spark Version: 3.5.0


### Additional context

_No response_

### Impacted by this bug?

Give it a 👍 We prioritize the issues with most 👍

--- a/pkg/util/sparkapplication.go
+++ b/pkg/util/sparkapplication.go
@@ -465,8 +465,10 @@ func GetExecutorRequestResource(app *v1beta2.SparkApplication) corev1.ResourceLi
 		}
 	}
 
+	instances := GetInitialExecutorNumber(app)
+
 	resourceList := []corev1.ResourceList{{}}
-	for i := int32(0); i < *app.Spec.Executor.Instances; i++ {
+	for i := int32(0); i < instances; i++ {
 		resourceList = append(resourceList, minResource)
 	}
 	return SumResourceList(resourceList)

--- a/pkg/util/sparkapplication_test.go
+++ b/pkg/util/sparkapplication_test.go
@@ -423,3 +423,58 @@ var _ = Describe("Check if IsDynamicAllocationEnabled", func() {
 		})
 	})
 })
+
+var _ = Describe("GetExecutorRequestResource", func() {
+	Context("when Executor.Instances is nil (dynamic allocation enabled)", func() {
+		app := &v1beta2.SparkApplication{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-app",
+				Namespace: "test-namespace",
+			},
+			Spec: v1beta2.SparkApplicationSpec{
+				DynamicAllocation: &v1beta2.DynamicAllocation{
+					Enabled:      true,
+					MinExecutors: ptr.To[int32](1),
+					MaxExecutors: ptr.To[int32](10),
+				},
+				Executor: v1beta2.ExecutorSpec{
+					SparkPodSpec: v1beta2.SparkPodSpec{
+						Cores:  ptr.To[int32](1),
+						Memory: ptr.To("1g"),
+					},
+				},
+			},
+		}
+
+		It("Should return resource list based on initial executor count without panicking", func() {
+			v1beta2.SetSparkApplicationDefaults(app)
+			Expect(app.Spec.Executor.Instances).To(BeNil())
+			Expect(func() { util.GetExecutorRequestResource(app) }).NotTo(Panic())
+			resources := util.GetExecutorRequestResource(app)
+			Expect(resources).NotTo(BeEmpty())
+		})
+	})
+
+	Context("when Executor.Instances is set", func() {
+		app := &v1beta2.SparkApplication{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-app",
+				Namespace: "test-namespace",
+			},
+			Spec: v1beta2.SparkApplicationSpec{
+				Executor: v1beta2.ExecutorSpec{
+					Instances: ptr.To[int32](2),
+					SparkPodSpec: v1beta2.SparkPodSpec{
+						Cores:  ptr.To[int32](1),
+						Memory: ptr.To("1g"),
+					},
+				},
+			},
+		}
+
+		It("Should return aggregated resources for all instances", func() {
+			resources := util.GetExecutorRequestResource(app)
+			Expect(resources).NotTo(BeEmpty())
+		})
+	})
+})

